src/scf/newton.jl

```diff
@@ -0,0 +1,420 @@
+# δφ[ik][:,i] = δφi for each k-point.
+# In this framework :
+#   - computing Ωδφ can be done analitically with the formula
+#         Ωδφi = H*δφi - Σj <δφj|H|δφi> δφj - Σj λji δφj
+#     where λij = <φi|H|φj>;
+#   - computing Kδφ can be done using the kernel function of DFTK after
+#     computing the associated δρ;
+#   - to perform Newton iterations, we solve (Ω+K)δφ = Hφ-λφ and then perform
+#     φ = φ - δφ, with proper orthonormalization.
+#
+# For further details :
+# Eric Cancès, Gaspard Kemlin, Antoine Levitt. Convergence analysis of
+# direct minimization and self-consistent iterations. SIAM Journal of Matrix
+# Analysis and Applications, 42(1):243–274 (2021)
+
+import KrylovKit: ArnoldiIterator, Orthogonalizer, OrthonormalBasis, KrylovDefaults, orthogonalize!
+using KrylovKit
+
+#
+# Tools
+#
+
+"""
+    compute_residual(basis::PlaneWaveBasis, φ, occ)
+
+Compute the residual associated to a set of planewave φ, that is to say
+H(φ)*φ - λ*φ where λ is the set of rayleigh coefficients associated to the φ.
+"""
+function compute_residual(basis::PlaneWaveBasis, φ, occ)
+
+    # necessary quantities
+    Nk = length(basis.kpoints)
+    ρ = compute_density(basis, φ, occ)
+    energies, H = energy_hamiltonian(basis, φ, occ; ρ=ρ)
+
+    # compute residual
+    res = similar(φ)
+    for ik = 1:Nk
+        φk = φ[ik]
+        N = size(φk, 2)
+        Hk = H.blocks[ik]
+        # eigenvalues as rayleigh coefficients
+        egvalk = φk'*(Hk*φk)
+        # compute residual at given kpoint as H(φ)φ - λφ
+        res[ik] = Hk*φk - φk*egvalk
+    end
+    res
+end
+
+# To project onto the space tangent to φ, we project δφ onto the orthogonal of φ
+function proj_tangent(δφ, φ; tol_test=1e-12)
+
+    Nk1 = size(δφ,1)
+    Nk2 = size(φ,1)
+    @assert Nk1 == Nk2
+    Nk = Nk1
+
+    Πδφ = similar(δφ)
+
+    for ik = 1:Nk
+        φk = φ[ik]
+        δφk = δφ[ik]
+        Πδφk = deepcopy(δφk)
+
+        N1 = size(δφk,2)
+        N2 = size(φk,2)
+        @assert N1 == N2
+        N = N1
+
+        for i = 1:N, j = 1:N
+            Πδφk[:,i] -= (φk[:,j]'δφk[:,i]) * φk[:,j]
+        end
+        Πδφ[ik] = Πδφk
+    end
+
+    # test orthogonalisation
+    for ik = 1:Nk
+        φk = φ[ik]
+        δφk = δφ[ik]
+        Πδφk = Πδφ[ik]
+        N = size(φk,2)
+        for i = 1:N, j = 1:N
+            @assert abs(Πδφk[:,i]'φk[:,j]) < tol_test [println(abs(Πδφk[:,i]'φk[:,j]))]
+        end
+    end
+
+    Πδφ
+end
+
+# KrylovKit custom orthogonaliser to be used in KrylovKit eigsolve, svdsolve,
+# linsolve, ...
+# This has to be passed in KrylovKit solvers when studying (Ω+K) as a super
+# operator from the tangent space to the tangent space to be sure that
+# iterations are constrained to stay on the tangent space.
+# /!\ Be careful that here, φ needs to be packed into on big array
+struct OrthogonalizeAndProject{F, O <: Orthogonalizer, φ} <: Orthogonalizer
+    projector::F
+    orth::O
+    φ::φ
```

Comment:
This is confusing. Use a different label for the type.

Reply:
I'm not sure what mean, can you explain ? :sweat_smile:

Reply:
phi::Tphi (not use the same identifier for the variable and the type)